### PR TITLE
Data Release Module

### PR DESCRIPTION
**Pull request by Jkat**
You can package your releases and distribute them here.

### CODE REVIEW

File: modules/data_release/ajax/FileUpload.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            array(
+             'file_name'   => $fileName,
+             'version'     => $version,
+             'upload_date' => $upload_date,
+            )
+        );
+    }
+    header("Location: ../main.php?test_name=data_release&uploadSuccess=true");
+} else {
+    echo "There was an error uploading the file";
```
Comment (driusan) [inline modules/data_release/ajax/FileUpload.php]: This should include a 400 Bad Request header too.

File: modules/data_release/ajax/GetFile.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *  @link     https://github.com/aces/Loris
+ */
+
+$user =& User::singleton();
+// Load config file and ensure paths are correct
+set_include_path(
+    get_include_path() . ":" .
+    __DIR__ . "/../project/libraries:" .
+    __DIR__ . "/../php/libraries"
+);
```
Comment (driusan) [inline modules/data_release/ajax/GetFile.php]: This looks like old code that should be replaced by requiring the autoload file.

File: modules/data_release/php/NDB_Menu_Filter_data_release.class.inc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        $this->columns      = array(
+                               'file_name',
+                               'version',
+                               'upload_date',
+                              );
+        $this->query        = " FROM data_release WHERE id IN"
+                              . " (SELECT data_release_id FROM"
+                              . " data_release_permissions WHERE"
+                              . " userid=(SELECT ID FROM users WHERE"
+                              . " UserID='{$user->getUsername()}'))";
```
Comment (driusan) [inline modules/data_release/php/NDB_Menu_Filter_data_release.class.inc]: This is a really awkward query.. can it be replaced by a join? Also, string concatentation in an SQL = injection attack.

File: modules/data_release/ajax/GetFile.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+$db         =& Database::singleton();
+$fileID     = $db->pselectOne(
+    "SELECT ID FROM data_release WHERE "
+    . "file_name=:fn",
+    array('fn' => $File)
+);
+$uid        = $db->pselectOne(
+    "SELECT ID FROM users WHERE "
+    . "UserID='{$user->getUsername()}'",
```
Comment (gluneau) [inline modules/data_release/ajax/GetFile.php]: This should be a prepared statement.

File: modules/data_release/ajax/GetFile.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    array()
+);
+$permission = $db->pselectOne(
+    "SELECT 'X' FROM data_release_permissions WHERE "
+    . "userid=$uid AND data_release_id=$fileID",
```
Comment (gluneau) [inline modules/data_release/ajax/GetFile.php]: this should be a prepared statement

**Comment by gluneau** [comment]
Please make note that deleting access or releases is not implemented.

File: modules/data_release/ajax/AddPermission.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    $userid          = $_POST['userid'];
+    $data_release_id = $_POST['data_release_id'];
+    $success         = $DB->insert(
+        'data_release_permissions',
+        array(
+         'userid'          => $userid,
+         'data_release_id' => $data_release_id,
+        )
+    );
+    header("Location: /data_release/?addpermissionSuccess=true");
```
Comment (driusan) [inline modules/data_release/ajax/AddPermission.php]: This is missing the baseURL so won't work on things not installed directly on the host @Jkat

File: modules/data_release/ajax/FileUpload.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        );
+        $success = $DB->insert(
+            'data_release_permissions',
+            array(
+             'userid'          => $user_ID,
+             'data_release_id' => $ID,
+            )
+        );
+    }
+    header("Location: /data_release/?uploadSuccess=true");
```
Comment (driusan) [inline modules/data_release/ajax/FileUpload.php]: same thing here.

**Comment by Jkat** [comment]
Thanks @driusan, fixed.

**Comment by driusan** [comment]
FILE: ...vis/build/aces/Loris/modules/data_release/ajax/AddPermission.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 27 | ERROR | [x] Whitespace found at end of line
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------

File: modules/data_release/php/NDB_Menu_Filter_data_release.class.inc [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        $this->columns      = array(
+                               'file_name',
+                               'version',
+                               'upload_date',
+                              );
+        $this->query        = " FROM data_release dr JOIN"
+                               . " data_release_permissions drp"
+                               . " ON (dr.id=drp.data_release_id)"
+                               . " JOIN users u ON (u.ID=drp.userid)"
+                               . " WHERE u.UserID='{$user->getUsername()}'";
```
Comment (xlecours) [inline modules/data_release/php/NDB_Menu_Filter_data_release.class.inc]: use DBquote

**Pull request merged**